src/checkers/inference/solver/GeneralSolver.java

```diff
@@ -0,0 +1,411 @@
+import checkers.inference.model.ConstantSlot;
+import checkers.inference.model.Constraint;
+import checkers.inference.model.Serializer;
+import checkers.inference.model.Slot;
+import checkers.inference.model.VariableSlot;
+import checkers.inference.solver.backend.BackEnd;
+import checkers.inference.solver.constraintgraph.ConstraintGraph;
+import checkers.inference.solver.constraintgraph.GraphBuilder;
+import checkers.inference.solver.frontend.ConstraintSerializer;
+import checkers.inference.solver.frontend.Lattice;
+import checkers.inference.solver.frontend.TwoQualifiersLattice;
+import checkers.inference.solver.util.Constants;
+import checkers.inference.solver.util.PrintUtils;
+import checkers.inference.solver.util.StatisticRecorder;
+import checkers.inference.solver.util.StatisticRecorder.StatisticKey;
+
+/**
+ * GeneralSolver is the entry point of general solver framework, and it is also
+ * the front end of whole solver system. GeneralSolver configures command line
+ * arguments, creates corresponding back end(s) and serializer, invokes the back
+ * end(s) and returns the solution.
+ * 
+ * @author jianchu
+ *
+ */
+
+public class GeneralSolver implements InferenceSolver {
+
+    protected String backEndType;
+    protected boolean useGraph;
+    private boolean solveInParallel;
+    private boolean collectStatistic;
+    private Lattice lattice;
+    private ConstraintGraph constraintGraph;
+    private BackEnd<?, ?> realBackEnd;
+
+    // Timing variables:
+    private long graphBuildingStart;
+    private long graphBuildingEnd;
+    private long solvingStart;
+    private long solvingEnd;
+
+    @Override
+    public InferenceSolution solve(Map<String, String> configuration, Collection<Slot> slots,
+            Collection<Constraint> constraints, QualifierHierarchy qualHierarchy,
+            ProcessingEnvironment processingEnvironment) {
+
+        InferenceSolution solution = null;
+
+        configure(configuration);
+        configureLattice(qualHierarchy);
+        Serializer<?, ?> defaultSerializer = createSerializer(backEndType, lattice);
+
+        if (useGraph) {
+            this.graphBuildingStart = System.currentTimeMillis();
+            this.constraintGraph = generateGraph(slots, constraints, processingEnvironment);
+            this.graphBuildingEnd = System.currentTimeMillis();
+            StatisticRecorder.record(StatisticKey.GRAPH_GENERATION_TIME, (graphBuildingEnd - graphBuildingStart));
+            solution = graphSolve(constraintGraph, configuration, slots, constraints, qualHierarchy,
+                    processingEnvironment, defaultSerializer);
+        } else {
+            realBackEnd = createBackEnd(backEndType, configuration, slots, constraints, qualHierarchy,
+                    processingEnvironment, lattice, defaultSerializer);
+            solution = solve();
+        }
+
+        if (solution == null) {
+            ErrorReporter.errorAbort("Null solution detected!");
```

Comment:
Well, If the solution is null here, there has to be something wrong in GTIS. It's different than constraint unsolvable.